test/ng/parseSpec.js

```diff
@@ -2973,6 +2973,111 @@ describe('parser', function() {
+          var filterCalls = 0;
+          $filterProvider.register('foo', valueFn(function(input) {
+            filterCalls++;
+            return input;
+          }));
+
+          scope.date = new Date(1234567890123);
+
+          var watcherCalls = 0;
+          scope.$watch('[(date | foo)]', function(input) {
+            watcherCalls++;
+          });
+
+          scope.$digest();
+          expect(filterCalls).toBe(1);
+          expect(watcherCalls).toBe(1);
+
+          scope.$digest();
+          expect(filterCalls).toBe(1);
+          expect(watcherCalls).toBe(1);
+        }));
+
+        it('should reevaluate filters in literals with non-primitive input that does support valueOf() when' +
+           ' valueOf() value changes',
+            inject(function($parse) {
+          var filterCalls = 0;
+          $filterProvider.register('foo', valueFn(function(input) {
+            filterCalls++;
+            return input;
+          }));
+
+          scope.date = new Date(1234567890123);
+
+          var watcherCalls = 0;
+          scope.$watch('[(date | foo)]', function(input) {
+            watcherCalls++;
+          });
+
+          scope.$digest();
+          expect(filterCalls).toBe(1);
+          expect(watcherCalls).toBe(1);
+
+          scope.date.setTime(1234567890);
+
+          scope.$digest();
+          expect(filterCalls).toBe(2);
+          expect(watcherCalls).toBe(2);
+        }));
+
+        it('should not reevaluate literals containing filters with non-primitive input that does support valueOf()' +
+            ' when the instance changes but valueOf() does not', inject(function($parse) {
+          var filterCalls = 0;
+          $filterProvider.register('foo', valueFn(function(input) {
+            filterCalls++;
+            return input;
+          }));
+
+          scope.date = new Date(1234567890123);
+
+          var watcherCalls = 0;
+          scope.$watch($parse('[(date | foo)]'), function(input) {
+            watcherCalls++;
+          });
+
+          scope.$digest();
+          expect(watcherCalls).toBe(1);
+
+          scope.date = new Date(1234567890123);
+          scope.$digest();
+          expect(watcherCalls).toBe(1);
+        }));
+
+        it('should not reevaluate filters with literal input containing primitives', inject(function($parse) {
+          var filterCalls = 0;
+          $filterProvider.register('foo', valueFn(function(input) {
+            filterCalls++;
+            return input;
+          }));
+
+          var watcherCalls = 0;
+          scope.$watch('[a] | foo', function(input) {
+            watcherCalls++;
+          });
+
+          scope.$apply("a = 1");
+          expect(filterCalls).toBe(1);
+          expect(watcherCalls).toBe(1);
+
+          scope.$apply("a = 2");
+          expect(filterCalls).toBe(2);
+          expect(watcherCalls).toBe(2);
+        }));
+
+        it('should always reevaluate filters with literal input containing non-primitives',
```

Comment:
One way to test this better (which I'll add) would be something like `[(obj | filterThatConvertsToPrimitive)]`